x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/pivot/CompositeBucketsChangeCollector.java

```diff
@@ -523,13 +707,22 @@ public boolean isOptimized() {
         return fieldCollectors.values().stream().anyMatch(FieldCollector::isOptimized);
     }
 
+    @Override
+    public boolean queryForChanges() {
+        return fieldCollectors.values().stream().anyMatch(FieldCollector::queryForChanges);
+    }
+
     public static ChangeCollector buildChangeCollector(
-        @Nullable CompositeAggregationBuilder compositeAggregationBuilder,
+        CompositeAggregationBuilder compositeAggregationBuilder,
         Map<String, SingleGroupSource> groups,
         String synchronizationField
     ) {
         Map<String, FieldCollector> fieldCollectors = createFieldCollectors(groups, synchronizationField);
-        return new CompositeBucketsChangeCollector(compositeAggregationBuilder, fieldCollectors);
+
+        // after building the field collectors we know whether we need the composite aggregation
+        boolean keepCompositeAggregatonBuilder = fieldCollectors.values().stream().anyMatch(FieldCollector::requiresCompositeSources);
+
+        return new CompositeBucketsChangeCollector(keepCompositeAggregatonBuilder ? compositeAggregationBuilder : null, fieldCollectors);
```

Comment:
This is a strange optimization to me. If you only want to optionally construct the `compositeAggregationBuilder` make it a `CachedSupplier` that is only called once or something.

Reply:
true, however I am not convinced a CachedSupplier is necessary. It makes it more complicated than necessary. Note that this is constructed exactly once after `_start`, so performance does not matter.